File: src/server/systems/units/can-move.ts
import { AnyEntity, useDeltaTime, useEvent, World } from "@rbxts/matter";
import * as Components from "shared/components";
import Remotes from "shared/remotes";

const moveUnitRemote = Remotes.Server.Get("MoveUnits");

/**
 * @param world Matter
 */
function canMove(world: World): void {
	Components.Health;
	for (const [, player, position, selectedUnits] of useEvent(moveUnitRemote, moveUnitRemote)) {
		print("Received move request");
		print(position);
		const x = position.X;
		const z = position.Y;

		selectedUnits.forEach((id) => {
			const unit = world.get(id as AnyEntity, Components.Unit);
			if (unit?.owner !== player) {
				warn("Received units not owned by sender");
				return;
			}

			world.insert(
				id as AnyEntity,
				Components.Target({
					location: new Vector3(x, 2, z),
				}),
			);
		});
	}

	for (const [id, target, transform, movement] of world.query(
		Components.Target,
		Components.Transform,
		Components.Movement,
		Components.Unit,
	)) {
		const direction = target.location.sub(transform.position).Unit;
		const newLocation = transform.position.add(direction.mul(movement.movementSpeed).mul(useDeltaTime()));

		world.insert(id, transform.patch({ position: newLocation }));

		if (target.location.sub(transform.position).Magnitude <= 0.1) {
			world.remove(id, Components.Target);
		}
	}
}

export = canMove;

File: old-src/server/Systems/Units/CanMove.lua
local ReplicatedStorage = game:GetService("ReplicatedStorage")

local Packages = ReplicatedStorage.Packages
local Matter = require(Packages.Matter)

local Components = require(ReplicatedStorage.Shared.Components)

local MoveUnitsRemote = ReplicatedStorage.MoveUnitsRemote

type MovementComponent = Components.MovementComponent
type TargetComponent = Components.TargetComponent
type TransformComponent = Components.TransformComponent
type UnitComponent = Components.UnitComponent
type World = Matter.World

local function CanMove(world: World)
    -- This may not want to be in this?
    for _, sender, position, selectedUnits in Matter.useEvent(MoveUnitsRemote, MoveUnitsRemote.OnServerEvent) do
        local x = position.X
        local z = position.Y

        for _, id in selectedUnits do
            local unit: UnitComponent = world:get(id, Components.Unit)
            if unit.owner ~= sender then
                warn("Received units not owned by sender")
                return
            end

            world:insert(
                id,
                Components.Target({
                    location = Vector3.new(x, 2, z),
                })
            )
        end
    end

    for id: number, target: TargetComponent, transform: TransformComponent, movement: MovementComponent in
        world:query(Components.Target, Components.Transform, Components.Movement, Components.Unit)
    do
        local direction = (target.location - transform.position).unit
        local newLocation = transform.position + direction * movement.movementSpeed * Matter.useDeltaTime()

        world:insert(
            id,
            transform:patch({
                position = newLocation,
            })
        )

        if (target.location - transform.position).Magnitude < 0.1 then
            world:remove(id, Components.Target)
        end
    end
end

return CanMove

--[=[
    User clicks on the ground. The player then needs to replicate the position of the location to the server along with the selected components.

    Could the troops have a target component that is replicated client > server?
]=]
